lightning/src/onion_message/router.rs

```diff
@@ -0,0 +1,248 @@
+) -> Result<Vec<PublicKey>, LightningError> where L::Target: Logger, GL::Target: Logger
+{
+	let graph_lock = network_graph.read_only();
+	let network_channels = graph_lock.channels();
+	let network_nodes = graph_lock.nodes();
+	let our_node_id = NodeId::from_pubkey(our_node_pubkey);
+
+	let mut for_each_successor = |node_id, callback: &mut FnMut(&NodeId, u64)| {
+		// TODO: in this method, check if OM forwarding feature bit is supported
+		if node_id == our_node_id && first_hops.is_some() {
+			if let Some(first_hops) = first_hops {
+				for hop in first_hops {
+					callback(&NodeId::from_pubkey(hop), 1);
+				}
+			}
+		} else if let Some(node_info) = network_nodes.get(&node_id) {
+			for scid in &node_info.channels {
+				if let Some(chan_info) = network_channels.get(&scid) {
+					let successor_node_id = if chan_info.node_one == node_id {
+						&chan_info.node_two
+					} else {
+						debug_assert!(chan_info.node_two == node_id);
+						&chan_info.node_one
+					};
+					callback(successor_node_id, 1); // Use a fixed cost for each hop until scoring is added
+				}
+			}
+		}
+	};
+
+	let mut invalid_final_hop_pk = None;
+	let mut convert_final_hop = |node_id: &NodeId| {
+		match PublicKey::from_slice(node_id.as_slice()) {
+			Ok(pk) => Ok(pk),
+			Err(e) => {
+				invalid_final_hop_pk = Some(*node_id);
+				Err(())
+			},
+		}
+	};
+
+	let receiver_node_id = NodeId::from_pubkey(receiver_pubkey);
+	match dijkstra(our_node_id, &mut for_each_successor, |node_id| node_id == &receiver_node_id, &mut convert_final_hop) {
+		Ok(p) => Ok(p),
+		Err(Error::PathNotFound) => Err(LightningError {
+			err: "Failed to find a path to the given destination".to_owned(),
+			action: ErrorAction::IgnoreError,
+		}),
+		Err(Error::FinalHopConversion) => {
+			debug_assert!(invalid_final_hop_pk.is_some());
+			Err(LightningError {
+				err: format!("Public key {:?} is invalid", invalid_final_hop_pk),
+				action: ErrorAction::IgnoreAndLog(Level::Trace)
+			})
+		}
+	}
+}
+
+#[derive(Debug, PartialEq)]
+/// Errored running `dijkstra`.
+enum Error {
+	/// No path exists to the destination.
+	PathNotFound,
+	/// Converting the processing hop type to the final hop type failed, see `dijkstra`'s
+	/// `convert_final_hop` parameter.
+	FinalHopConversion,
+}
+
+//  Heavily adapted from https://github.com/samueltardieu/pathfinding/blob/master/src/directed/dijkstra.rs
+//  TODO: how2credit the repo (is that necessary?)?
+/// Run Dijkstra's from `start` until `found_target` indicates that we've found the destination.
+/// `successor_callback` must invoke the callback that it is provided on each of a given node's
+/// next-hop peers. `convert_final_hop` may be used to convert an intermediate processing hop type
+/// (`N`) to a final path hop type (`H`).
+fn dijkstra<N, H, FN, FS, FC>(start: N, successor_callback: &mut FN, found_target: FS,
```

Comment:
This interface seems to be kept very generic to accommodate different usage scenarios by the original author. I think when we're adapting this we can fill in a lot of these types, which should improve readability?